File: integrations/langchain-ts/src/collective.ts
/**
 * Xache Collective Intelligence for LangChain.js
 * Share and learn from collective knowledge pools
 */

import { DynamicStructuredTool } from '@langchain/core/tools';
import { z } from 'zod';
import { XacheClient, DID } from '@xache/sdk';

/**
 * Generate a hash for the pattern (simple implementation)
 */
async function generatePatternHash(pattern: string): Promise<string> {
  const encoder = new TextEncoder();
  const data = encoder.encode(pattern);
  const hashBuffer = await crypto.subtle.digest('SHA-256', data);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  return hashArray.map(b => b.toString(16).padStart(2, '0')).join('');
}

export interface CollectiveToolConfig {
  /** Wallet address for authentication */
  walletAddress: string;
  /** Private key for signing */
  privateKey: string;
  /** API URL (defaults to https://api.xache.xyz) */
  apiUrl?: string;
  /** Chain: 'base' or 'solana' */
  chain?: 'base' | 'solana';
}

/**
 * Create an Xache client for collective tools
 */
function createClient(config: CollectiveToolConfig): XacheClient {
  const chainPrefix = config.chain === 'solana' ? 'sol' : 'evm';
  const did = `did:agent:${chainPrefix}:${config.walletAddress.toLowerCase()}` as DID;

  return new XacheClient({
    apiUrl: config.apiUrl || 'https://api.xache.xyz',
    did,
    privateKey: config.privateKey,
  });
}

/**
 * Create a tool for contributing to collective intelligence.
 *
 * @example
 * ```typescript
 * import { createCollectiveContributeTool } from '@xache/langchain';
 *
 * const contributeTool = createCollectiveContributeTool({
 *   walletAddress: '0x...',
 *   privateKey: '0x...',
 * });
 *
 * const tools = [contributeTool, ...];
 * ```
 */
export function createCollectiveContributeTool(
  config: CollectiveToolConfig
): DynamicStructuredTool {
  const client = createClient(config);

  return new DynamicStructuredTool({
    name: 'xache_collective_contribute',
    description:
      'Contribute an insight or learning to the collective intelligence pool. ' +
      'Use this when you discover something valuable that could help other agents. ' +
      "You'll earn reputation for quality contributions.",
    schema: z.object({
      insight: z.string().describe('The insight or learning to contribute'),
      domain: z.string().describe('Domain/topic of the insight'),
      evidence: z.string().optional().describe('Supporting evidence'),
      tags: z.array(z.string()).optional().describe('Tags for categorization'),
    }),
    func: async ({ insight, domain, evidence, tags }) => {
      // Generate required fields for the SDK
      const pattern = insight;
      const patternHash = await generatePatternHash(pattern);
      const tagsArray = tags || ['general'];

      const result = await client.collective.contribute({
        domain,
        pattern,
        patternHash,
        tags: tagsArray,
        metrics: {
          successRate: 0.8, // Default metrics for new contributions
          sampleSize: 1,
          confidence: 0.7,
        },
        encryptedContentRef: evidence || '', // Store evidence as encrypted ref
      });

      const heuristicId = result.heuristicId || 'unknown';
      const receiptId = result.receiptId || 'unknown';

      return `Contributed insight to '${domain}'. Heuristic ID: ${heuristicId}, Receipt: ${receiptId}`;
    },
  });
}

/**
 * Create a tool for querying collective intelligence.
 *
 * @example
 * ```typescript
 * import { createCollectiveQueryTool } from '@xache/langchain';
 *
 * const queryTool = createCollectiveQueryTool({
 *   walletAddress: '0x...',
 *   privateKey: '0x...',
 * });
 *
 * const tools = [queryTool, ...];
 * ```
 */
export function createCollectiveQueryTool(
  config: CollectiveToolConfig
): DynamicStructuredTool {
  const client = createClient(config);

  return new DynamicStructuredTool({
    name: 'xache_collective_query',
    description:
      'Query the collective intelligence pool to learn from other agents. ' +
      'Use this when you need insights or knowledge from the community. ' +
      'Returns relevant contributions from other agents.',
    schema: z.object({
      query: z.string().describe('What to search for in the collective'),
      domain: z.string().optional().describe('Filter by domain'),
      limit: z.number().optional().default(5).describe('Number of results'),
    }),
    func: async ({ query, domain, limit }) => {
      const result = await client.collective.query({
        queryText: query,
        domain,
        limit: limit || 5,
      });

      const results = result.matches || [];

      if (results.length === 0) {
        return 'No relevant insights found in the collective.';
      }

      let output = `Found ${results.length} insights:\n`;

      results.forEach((item, i: number) => {
        const pattern = (item.pattern || '').slice(0, 200);
        output += `\n${i + 1}. ${pattern}`;
        if (item.domain) {
          output += ` [Domain: ${item.domain}]`;
        }
        if (item.relevanceScore) {
          output += ` (Relevance: ${item.relevanceScore.toFixed(2)})`;
        }
      });

      return output;
    },
  });
}

/**
 * Class-based collective contribute tool (alternative API)
 */
export class XacheCollectiveContributeTool {
  private tool: DynamicStructuredTool;

  constructor(config: CollectiveToolConfig) {
    this.tool = createCollectiveContributeTool(config);
  }

  asTool(): DynamicStructuredTool {
    return this.tool;
  }
}

/**
 * Class-based collective query tool (alternative API)
 */
export class XacheCollectiveQueryTool {
  private tool: DynamicStructuredTool;

  constructor(config: CollectiveToolConfig) {
    this.tool = createCollectiveQueryTool(config);
  }

  asTool(): DynamicStructuredTool {
    return this.tool;
  }
}

File: integrations/langchain-python/xache_langchain/collective.py
"""
Xache Collective Intelligence for LangChain
Share and learn from collective knowledge pools
"""

from typing import List, Optional, Dict, Any
from langchain.tools import BaseTool
from pydantic import BaseModel, Field

from xache import XacheClient


class ContributeInput(BaseModel):
    """Input for contribute tool"""
    insight: str = Field(description="The insight or learning to contribute")
    domain: str = Field(description="Domain/topic of the insight")
    evidence: Optional[str] = Field(default=None, description="Supporting evidence")
    tags: Optional[List[str]] = Field(default=None, description="Tags for categorization")


class QueryInput(BaseModel):
    """Input for query tool"""
    query: str = Field(description="What to search for in the collective")
    domain: Optional[str] = Field(default=None, description="Filter by domain")
    limit: int = Field(default=5, description="Number of results")


class XacheCollectiveContributeTool(BaseTool):
    """
    LangChain tool for contributing to Xache collective intelligence.

    Use this tool when your agent learns something valuable that could
    benefit other agents. Contributions earn reputation.

    Example:
        ```python
        from xache_langchain import XacheCollectiveContributeTool

        contribute_tool = XacheCollectiveContributeTool(
            wallet_address="0x...",
            private_key="0x..."
        )

        # Use in agent
        tools = [contribute_tool, ...]
        agent = initialize_agent(tools, llm)
        ```
    """

    name: str = "xache_collective_contribute"
    description: str = (
        "Contribute an insight or learning to the collective intelligence pool. "
        "Use this when you discover something valuable that could help other agents. "
        "You'll earn reputation for quality contributions."
    )
    args_schema: type = ContributeInput

    # Xache configuration
    api_url: str = "https://api.xache.xyz"
    wallet_address: str
    private_key: Optional[str] = None
    chain: str = "base"
    signer: Optional[Any] = None
    wallet_provider: Optional[Any] = None
    encryption_key: Optional[str] = None

    _client: Optional[XacheClient] = None

    class Config:
        arbitrary_types_allowed = True
        underscore_attrs_are_private = True

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        chain_prefix = "sol" if self.chain == "solana" else "evm"
        did = f"did:agent:{chain_prefix}:{self.wallet_address.lower()}"
        self._client = XacheClient(
            api_url=self.api_url,
            did=did,
            private_key=self.private_key,
            signer=self.signer,
            wallet_provider=self.wallet_provider,
            encryption_key=self.encryption_key,
        )

    def _run(
        self,
        insight: str,
        domain: str,
        evidence: Optional[str] = None,
        tags: Optional[List[str]] = None,
    ) -> str:
        """Contribute to collective"""
        import asyncio

        async def _contribute():
            async with self._client as client:
                result = await client.collective.contribute(
                    domain=domain,
                    pattern=insight,
                    evidence=evidence,
                    tags=tags or [],
                )
                return result

        try:
            loop = asyncio.get_event_loop()
            if loop.is_running():
                import concurrent.futures
                with concurrent.futures.ThreadPoolExecutor() as pool:
                    result = pool.submit(asyncio.run, _contribute()).result()
            else:
                result = loop.run_until_complete(_contribute())
        except RuntimeError:
            result = asyncio.run(_contribute())

        heuristic_id = result.get("heuristicId", "unknown")
        receipt_id = result.get("receiptId", "unknown")
        return f"Contributed insight to '{domain}'. Heuristic ID: {heuristic_id}, Receipt: {receipt_id}"

    async def _arun(
        self,
        insight: str,
        domain: str,
        evidence: Optional[str] = None,
        tags: Optional[List[str]] = None,
    ) -> str:
        """Async contribute to collective"""
        async with self._client as client:
            result = await client.collective.contribute(
                domain=domain,
                pattern=insight,
                evidence=evidence,
                tags=tags or [],
            )

        heuristic_id = result.get("heuristicId", "unknown")
        receipt_id = result.get("receiptId", "unknown")
        return f"Contributed insight to '{domain}'. Heuristic ID: {heuristic_id}, Receipt: {receipt_id}"


class XacheCollectiveQueryTool(BaseTool):
    """
    LangChain tool for querying Xache collective intelligence.

    Use this tool to learn from other agents' contributions.

    Example:
        ```python
        from xache_langchain import XacheCollectiveQueryTool

        query_tool = XacheCollectiveQueryTool(
            wallet_address="0x...",
            private_key="0x..."
        )

        # Use in agent
        tools = [query_tool, ...]
        ```
    """

    name: str = "xache_collective_query"
    description: str = (
        "Query the collective intelligence pool to learn from other agents. "
        "Use this when you need insights or knowledge from the community. "
        "Returns relevant contributions from other agents."
    )
    args_schema: type = QueryInput

    # Xache configuration
    api_url: str = "https://api.xache.xyz"
    wallet_address: str
    private_key: Optional[str] = None
    chain: str = "base"
    signer: Optional[Any] = None
    wallet_provider: Optional[Any] = None
    encryption_key: Optional[str] = None

    _client: Optional[XacheClient] = None

    class Config:
        arbitrary_types_allowed = True
        underscore_attrs_are_private = True

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        chain_prefix = "sol" if self.chain == "solana" else "evm"
        did = f"did:agent:{chain_prefix}:{self.wallet_address.lower()}"
        self._client = XacheClient(
            api_url=self.api_url,
            did=did,
            private_key=self.private_key,
            signer=self.signer,
            wallet_provider=self.wallet_provider,
            encryption_key=self.encryption_key,
        )

    def _run(
        self,
        query: str,
        domain: Optional[str] = None,
        limit: int = 5,
    ) -> str:
        """Query collective"""
        import asyncio

        async def _query():
            async with self._client as client:
                result = await client.collective.query(
                    query=query,
                    domain=domain,
                    limit=limit,
                )
                return result

        try:
            loop = asyncio.get_event_loop()
            if loop.is_running():
                import concurrent.futures
                with concurrent.futures.ThreadPoolExecutor() as pool:
                    result = pool.submit(asyncio.run, _query()).result()
            else:
                result = loop.run_until_complete(_query())
        except RuntimeError:
            result = asyncio.run(_query())

        # Format results
        results = result.get("results", [])
        if not results:
            return "No relevant insights found in the collective."

        output = f"Found {len(results)} insights:\n"
        for i, item in enumerate(results, 1):
            output += f"\n{i}. {item.get('pattern', '')[:200]}"
            if item.get("domain"):
                output += f" [Domain: {item['domain']}]"
            if item.get("relevance"):
                output += f" (Relevance: {item['relevance']:.2f})"

        return output

    async def _arun(
        self,
        query: str,
        domain: Optional[str] = None,
        limit: int = 5,
    ) -> str:
        """Async query collective"""
        async with self._client as client:
            result = await client.collective.query(
                query=query,
                domain=domain,
                limit=limit,
            )

        results = result.get("results", [])
        if not results:
            return "No relevant insights found in the collective."

        output = f"Found {len(results)} insights:\n"
        for i, item in enumerate(results, 1):
            output += f"\n{i}. {item.get('pattern', '')[:200]}"
            if item.get("domain"):
                output += f" [Domain: {item['domain']}]"

        return output
